### PR TITLE
Us39985

### PR DESCRIPTION
**Pull request by FrankGasparovic**
FrankGasparovic US39985:
- Added resource consistency tests
- Added subject consistency tests

Signed-off-by: Matias matias.altman@ge.com

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -67,7 +67,7 @@
         <flyway.version>3.0</flyway.version>
         <spring-security-oauth2.version>2.0.3.RELEASE</spring-security-oauth2.version>
         <spring-security-jwt.version>1.0.2.RELEASE</spring-security-jwt.version>
-        <titandb.version>1.0.0</titandb.version>
+        <titandb.version>1.1.0-SNAPSHOT</titandb.version>
```
Comment (sanjeevchopra) [inline pom.xml:70]: these changes are already in develop ? please rebase to develop and squash commits in the branch

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceConsistencyTest.java [hunk 2]
```diff
@@ -37,25 +51,98 @@ private void setupTitanGraph() throws InterruptedException, ExecutionException {
         GraphConfig.createSchemaElements(this.graph);
     }
 
-    @Test(/* threadPoolSize = 2, invocationCount = 2, */ dataProvider = "resourcesForTestConcurrent",
+    @Test(
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceConsistencyTest.java]: Please make sure that acs team preferences are imported into eclipse. see eclipse-config/README.md

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceConsistencyTest.java [hunk 3]
```diff
@@ -37,25 +51,98 @@ private void setupTitanGraph() throws InterruptedException, ExecutionException {
...
+            /* threadPoolSize = 2, invocationCount = 2, */ dataProvider = "resourcesForTestConcurrent",
             successPercentage = 97)
-    public void testConcurrentWriteByZoneAndResourceIdentifier(final String resourceId) {
-        System.out
-                .println("thread_id: " + Long.toString(Thread.currentThread().getId()) + ", resource_id:" + resourceId);
+    public void testWriteByZoneAndResourceIdentifierConcurrent(final String resourceId)
+            throws PermanentLockingException {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceConsistencyTest.java]: why is this needed ?

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceConsistencyTest.java [hunk 4]
```diff
@@ -37,25 +51,98 @@ private void setupTitanGraph() throws InterruptedException, ExecutionException {
...
+    public void testDeleteConcurrent() {
+        ResourceEntity resourceEntity = persistResource1toZone1AndAssert();
+        assertThat(this.resourceRepository.findOne(resourceEntity.getId()), equalTo(resourceEntity));
+        this.resourceRepository.delete(resourceEntity.getId());
+        assertThat(this.resourceRepository.findOne(resourceEntity.getId()), nullValue());
+
+    }
+
+    @Test(threadPoolSize = 2, invocationCount = 100)
+    public void testFindAllConcurrent() throws PermanentLockingException {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceConsistencyTest.java]: this comment applies to all tests:

This test is expected to fail for some invocations because of PermanentLockingException (unique index lock).  You could handle the exception, though you cannot put a threshold on how many calls can fail.

I would use successPercentage - its better to ensure that no more than 3% tests fail even if we dont track which specific failure.

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceConsistencyTest.java [hunk 5]
```diff
@@ -37,25 +51,98 @@ private void setupTitanGraph() throws InterruptedException, ExecutionException {
...
+        assertThat(this.resourceRepository.findOne(resourceEntity1.getId()), equalTo(resourceEntity1));
+        ResourceEntity resourceEntity2 = persistResource2toZone1AndAssert();
+        assertThat(this.resourceRepository.findOne(resourceEntity2.getId()), equalTo(resourceEntity2));
+
+        List<ResourceEntity> resources = this.resourceRepository.findByZone(TEST_ZONE_1);
+        assertThat(resources, hasItems(resourceEntity1, resourceEntity2));
+    }
+
+    @Test(threadPoolSize = 2, invocationCount = 100)
+    public void testSaveConcurrent() throws PermanentLockingException {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceConsistencyTest.java]: this test doestn have any new coverage.

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceConsistencyTest.java [hunk 6]
```diff
@@ -37,25 +51,98 @@ private void setupTitanGraph() throws InterruptedException, ExecutionException {
...
+        assertThat(this.resourceRepository.findOne(resourceEntity.getId()), equalTo(resourceEntity));
+        assertThat(this.resourceRepository.findOne(resourceEntity.getId()).getAttributesAsJson(),
+                equalTo(resourceEntity.getAttributesAsJson()));
+
+    }
+
     @DataProvider(parallel = true)
     public Object[][] resourcesForTestConcurrent() {
-        return IntStream.range(0, 100).mapToObj(i -> new Object[] { "r" + Integer.toString(i) })
+        Object[][] test = IntStream.range(0, 100).mapToObj(i -> new Object[] { "r" + Integer.toString(i) })
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceConsistencyTest.java]: why create the variable ?

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectConsistencyTest.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+import com.ge.predix.acs.config.GraphConfig;
+import com.ge.predix.acs.model.Attribute;
+import com.ge.predix.acs.rest.Parent;
+import com.ge.predix.acs.utils.JsonUtils;
+import com.ge.predix.acs.zone.management.dao.ZoneEntity;
+import com.thinkaurelius.titan.core.TitanFactory;
+import com.thinkaurelius.titan.diskstorage.locking.PermanentLockingException;
+
+public class GraphSubjectConsistencyTest {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectConsistencyTest.java]: This and GraphResourceConsistencyTest are functionally the same, except they add concurrency ?   Did you consider, modifying the existing tests or reusing the code ?   As is, these tests may be hard to maintain due to a high degree of overlap.  lets discuss.

**Review by sanjeevchopra — CHANGES_REQUESTED** [review]

**Comment by FrankGasparovic** [comment]
Jenkins test are successful

[Build Status](https://predix1.jenkins.build.ge.com/job/Predix-Security/job/acs-build-n-test-pullrequest/687/)

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 8]
```diff
@@ -141,42 +142,27 @@ public void testFindAll() {
...
-        ResourceEntity resourceEntity1 = persist2LevelHierarchicalResource1toZone1();
-        assertThat(IteratorUtils.count(this.graph.vertices()), equalTo(2L));
-        persistResource1toZone2AndAssert();
-        assertThat(IteratorUtils.count(this.graph.vertices()), equalTo(3L));
+        ResourceEntity resourceEntity1 = persistRandomResourcetoZone1AndAssert();
 
         ResourceEntity resource = this.resourceRepository.getByZoneAndResourceIdentifier(TEST_ZONE_1,
                 resourceEntity1.getResourceIdentifier());
         assertThat(resource, equalTo(resourceEntity1));
-        // Check that the result does not contain inherited attribute
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java:168]: this check should be added in another test related to inheritance

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 9]
```diff
@@ -187,44 +173,39 @@ public void testGetByZoneAndResourceIdentifierWithInheritedAttributes() {
...
                 resourceIdentifier);
         assertThat(resource.getAttributes().contains(SITE_BASEMENT), equalTo(true));
         assertThat(resource.getAttributes().contains(TYPE_MONSTER_OF_THE_WEEK), equalTo(true));
     }
 
-    @Test
+    @Test(threadPoolSize = 3, invocationCount = 100)
     public void testGetByZoneAndResourceIdentifierWithInheritedAttributes3LevelHierarchical() {
-        assertThat(IteratorUtils.count(this.graph.vertices()), equalTo(0L));
         String resourceIdentifier = persist3LevelHierarchicalResource1toZone1().getResourceIdentifier();
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: rename persist3LevelHierarchicalResource1toZone1  to persist3LevelRandomResourcetoZone1

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 10]
```diff
@@ -235,6 +216,7 @@ public void testGetByZoneAndResourceIdentifierWithInheritedAttributes3LevelHiera
 
     @Test(expectedExceptions = SchemaViolationException.class)
     public void testPreventEntityParentSelfReference() {
+        this.dropAllResources();
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: this test should not require dropAllResources, if it uses a random resource id

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 11]
```diff
@@ -270,31 +253,23 @@ public void testPreventEntityParentCyclicReference() {
 
         resource1.setParents(
                 new HashSet<>(Arrays.asList(new Parent[] { new Parent(resource3.getResourceIdentifier()) })));
-        try {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java:273]: there is value in asserting the specific message. lets not remove it.

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 12]
```diff
@@ -270,31 +253,23 @@ public void testPreventEntityParentCyclicReference() {
...
-            assertThat(ex.getMessage(), equalTo(
-                    "Updating entity '/site/basement' with parent '/evidence/implant' introduces a cyclic reference."));
-            return;
-        }
-        Assert.fail("save() did not throw the expected exception.");
+        this.resourceRepository.save(resource1);
     }
 
-    @Test
+    @Test(threadPoolSize = 3, invocationCount = 100)
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: since we  have retry, we should reduce the invocationCount for all concurrent tests to save execution time. 50 ?  we can test a few numbers to see how much time we save.

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 13]
```diff
@@ -304,91 +279,93 @@ public void testSaveWithNoZoneName() {
 
     @Test(expectedExceptions = IllegalArgumentException.class)
     public void testSaveWithNoZoneEntity() {
+        this.dropAllResources();
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: not needed. dropAll slows down tests - we should limit to where it's needed.

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 14]
```diff
@@ -304,91 +279,93 @@ public void testSaveWithNoZoneName() {
...
-        String resourceId = resourceIdentifier;
         GraphTraversal<Vertex, Vertex> traversal = g.V().has(RESOURCE_ID_KEY, resourceId);
         assertThat(traversal.hasNext(), equalTo(true));
         assertThat(traversal.next().property(RESOURCE_ID_KEY).value(), equalTo(resourceId));
 
         // Update the resource.
-        resourceEntity.setAttributesAsJson("{\'status':'test'}");
-        this.resourceRepository.save(resourceEntity);
-        assertThat(IteratorUtils.count(this.graph.vertices()), equalTo(1L));
+        String updateJSON = "{\'status':'test'}";
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java:324]: minor: unnecessary variable?

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 15]
```diff
@@ -304,91 +279,93 @@ public void testSaveWithNoZoneName() {
...
+        resourceEntity.setAttributesAsJson(updateJSON);
+        concurrentSave(resourceEntity, 0);
+        assertThat(this.resourceRepository.getEntity(TEST_ZONE_1, resourceEntity.getResourceIdentifier())
+                .getAttributesAsJson(), equalTo(updateJSON));
     }
 
     @Test(expectedExceptions = SchemaViolationException.class)
     public void testUpdateUnattachedEntity() {
+        this.dropAllResources();
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: not needed

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 16]
```diff
@@ -304,91 +279,93 @@ public void testSaveWithNoZoneName() {
...
         assertThat(IteratorUtils.count(this.graph.vertices()), equalTo(0L));
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: remove count() anyway, even though this is being run in a single thread.

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 17]
```diff
@@ -304,91 +279,93 @@ public void testSaveWithNoZoneName() {
...
-        String resourceIdentifier = persistResource1toZone1AndAssert().getResourceIdentifier();
+        ResourceEntity resource = persistRandomResourcetoZone1AndAssert();
+        String resourceId = resource.getResourceIdentifier();
         assertThat(IteratorUtils.count(this.graph.vertices()), equalTo(1L));
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: rm

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 18]
```diff
@@ -304,91 +279,93 @@ public void testSaveWithNoZoneName() {
...
+    @Test(expectedExceptions = IllegalStateException.class)
     public void testSaveResourceWithNonexistentParent() {
+        this.dropAllResources();
         assertThat(IteratorUtils.count(this.graph.vertices()), equalTo(0L));
         ResourceEntity resource = new ResourceEntity(TEST_ZONE_1, DRIVE_ID);
         resource.setParents(new HashSet<>(Arrays.asList(new Parent(BASEMENT_SITE_ID))));
 
         // Save a resource with nonexistent parent which should throw IllegalStateException exception
         // while saving parent relationships.
-        try {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java:384]: lets not remove the assert for the message.

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 19]
```diff
@@ -494,32 +471,45 @@ public void testGetResourceEntityAndDescendantsIds() {
         assertThat(descendantsIds, empty());
     }
 
+    @Test(expectedExceptions = IllegalStateException.class)
+    public void testChildWithDifferentZone() {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java:482]: 👍

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 20]
```diff
@@ -494,32 +471,45 @@ public void testGetResourceEntityAndDescendantsIds() {
...
+        ResourceEntity parentResource = persistResourceToZoneAndAssert(TEST_ZONE_1,
+                BASEMENT_SITE_ID + getRandomNumber(), BASEMENT_ATTRIBUTES);
         HashSet<Parent> parents = new HashSet<>(
                 Arrays.asList(new Parent[] { new Parent(parentResource.getResourceIdentifier()) }));
-        return persistResourceWithParentsToZoneAndAssert(TEST_ZONE_1, DRIVE_ID, DRIVE_ATTRIBUTES, parents);
+        return persistResourceWithParentsToZoneAndAssert(TEST_ZONE_1, DRIVE_ID + getRandomNumber(), DRIVE_ATTRIBUTES,
+                parents);
     }
 
     public ResourceEntity persist3LevelHierarchicalResource1toZone1() {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: rename: persist3LevelRandomResourcetoZone1

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 21]
```diff
@@ -528,33 +518,53 @@ private ResourceEntity persistResourceWithParentsToZoneAndAssert(final ZoneEntit
...
         return persistResourceToZoneAndAssert(TEST_ZONE_1, JOSECHUNG_ID, DRIVE_ATTRIBUTES);
     }
 
-    private ResourceEntity persistResource1toZone2AndAssert() {
-        return persistResourceToZoneAndAssert(TEST_ZONE_2, DRIVE_ID, DRIVE_ATTRIBUTES);
+    private int getRandomNumber() {
+        return randomGenerator.nextInt(10000000);
+    }
+
+    private ResourceEntity concurrentSave(ResourceEntity resource, int retryTimes) throws TitanException {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: - rename to saveWithRetry ?  
- rename retryTimes to retryCount. honor retryCount by counting down to 0
- also  this method can be shared between Resource and Subject tests.  (use the base class ZonableEntity)

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 22]
```diff
@@ -528,33 +518,53 @@ private ResourceEntity persistResourceWithParentsToZoneAndAssert(final ZoneEntit
...
+        ResourceEntity resourceEntity = null;
+        try {
+            resourceEntity = this.resourceRepository.save(resource);
+        } catch (TitanException te) {
+            if (te.getCause().getCause().getMessage().contains("Local lock") && retryTimes <= 3) {
+                try {
+                    Thread.sleep(500);
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: try with a small sleep, 250ms ?

**Review by sanjeevchopra — CHANGES_REQUESTED** [review]
reviewed  GraphResourceRepositoryTest class

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java (hunk 14 above)
Comment (FrankGasparovic) [reply service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java:324]: Variable used twice, should we keep or just use the string values?

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java (hunk 21 above)
Comment (matiasa123) [reply service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: On the third bullet point are we talking about concurrentSave or getRandomNumber()?
